src/storage/tests/setup.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Unit tests for sources.
+
+use std::collections::{BTreeMap, HashMap};
+use std::marker::{Send, Sync};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+use timely::progress::{Antichain, Timestamp as _};
+
+use mz_build_info::DUMMY_BUILD_INFO;
+use mz_ore::metrics::MetricsRegistry;
+use mz_ore::now::SYSTEM_TIME;
+use mz_ore::task::RuntimeExt;
+use mz_repr::TimestampManipulation;
+use mz_repr::{Diff, GlobalId, Timestamp};
+use mz_storage::protocol::client::StorageCommand;
+use mz_storage::sink::SinkBaseMetrics;
+use mz_storage::source::metrics::SourceBaseMetrics;
+use mz_storage::source::testscript::ScriptCommand;
+use mz_storage::types::sources::{
+    encoding::SourceDataEncoding, SourceConnection, SourceData, SourceDesc, SourceEnvelope,
+    TestScriptSourceConnection,
+};
+use mz_storage::DecodeMetrics;
+
+pub fn assert_source_results_in(
+    source: Vec<ScriptCommand>,
+    encoding: SourceDataEncoding,
+    envelope: SourceEnvelope,
+    expected_values: Vec<SourceData>,
+) -> Result<(), anyhow::Error> {
+    let timestamp_interval = Duration::from_secs(1);
+
+    let desc = SourceDesc {
+        connection: SourceConnection::TestScript(TestScriptSourceConnection {
+            desc_json: serde_json::to_string(&source).unwrap(),
+        }),
+        encoding,
+        envelope,
+        metadata_columns: vec![],
+        timestamp_interval,
+    };
+
+    build_and_run_source(desc, timestamp_interval, move |upper, mut rh| {
+        let expected_values = expected_values.clone();
+        async move {
+            let as_of = if let Some(cur_time) = upper.as_option().copied() {
+                Antichain::from_elem(cur_time.step_back().unwrap_or_else(Timestamp::minimum))
+            } else {
+                // terminated source, we try to fetch everything
+                Antichain::from_elem(Timestamp::maximum())
+            };
+            let mut snapshot = rh.snapshot_and_fetch(as_of).await.unwrap();
+
+            // retry until we have enough data.
+            //
+            // TODO(guswynn): have a builtin timeout here for a source
+            // not producing enough data.
+            if snapshot.len() < expected_values.len() {
+                false
+            } else {
+                differential_dataflow::consolidation::consolidate_updates(&mut snapshot);
```

Comment:
its possible im using this wrong, and i should be using `consolidate` by ignoring the time component